### ISSUE TITLE
IE11 Invalid character Error 

### ISSUE THREAD

**Issue by azazlhr** [smart-on-fhir/client-js#55]
Hi,

I'm using fhir-client v0.1.15 to develop a SMART on FHIR web application that would run in EHR launch scenario. For example in Epic Hyperspace simulator that uses IE to run the SMART application.

While testing my application locally on IE11, I get a JavaScript error from within fhir-client.js. 

> SCRIPT1014: Invalid character
> File: fhir-client.js, Line: 38113, Column: 3

Below are the two lines from code. 38112 & 38113 in order:

```
  var ret = Adapter.get().defer();
  var state = JSON.parse(sessionStorage[params.state]);
```

Any solution or workaround would be much appreciated.

Post on [SO](https://stackoverflow.com/questions/57009966/smart-on-fhir-client-js-invalid-character-error-on-ie11) with bounty.

**Comment by vlad-ignatov** [comment]
Hi

The code itself does not seem to contain invalid characters. There two things that you can do:

1. This sounds like a parse error originating from that “JSON.parse”. Check what you have in sessionStorage and test if it can be parsed as JSON.
2. IE will sometimes report wrong line numbers. Try setting a debugger (pause on exceptions) to see if that will show more info.

Thanks,
Vlad

**Comment by azazlhr** [comment]
Thank you Vlad for quick response, 
Actually you are right about error originating from "JSON.parse". 
I've thoroughly debugged the code, and found that sessionStorage gets cleared in IE11 after redirection to authorization server. 
Code line:
`window.location.href = redirect_to; // line 38571 ` 

Does this imply that the library is not supported for IE11?

**Comment by vlad-ignatov** [comment]
No, that should be fine. The sessionStorage  is scoped per origin. When you get redirected to the authorization server, you are looking at another instance of sessionStorage that contains data (if any) assigned to the auth server's origin.

Thanks,
Vlad

**Comment by azazlhr** [comment]
Thanks Vlad, 

In fact, I've debugged fhir-client.js code, and observed two different behaviors for two different browsers 1. Chrome 2. IE11. 

In Chrome we always get what object from sessionStorage[state] that was set by the library before redirection. But in IE11, after redirect, we get undefined.

So basically everything is identical, library, scenario, flow but we get different results for IE.

**Comment by vlad-ignatov** [comment]
I think I lost you here :( Can you provide some example?

**Comment by azazlhr** [comment]
OK let me rephrase the whole scenario.
The issue is related to IE11 only.
My SMART on FHIR application calls FHIR.oauth2.authorize method of fhir-client.js library on its launch. 
For example when testing with AllScripts [EHR Launch](https://developer.allscripts.com/content/fhir/content/smart%20on%20fhir/), my launch URL is:

http://localhost:52447/AllScript/Launch?launch=54520&iss=https://pro171.open.allscripts.com/FHIR

Below is the code from my AllScripts/Launch.html page.
 ```
<script>
        FHIR.oauth2.authorize({
            "client": {                
                "client_id": "xxxxxxxx-xxxx-4b36-xxxx-551069175650",                 
                "scope": "launch user/*.* patient/*.read  openid fhirUser online_access",                
                "redirect_uri": "http://localhost:52447"                
            }            
        });
    </script>
```
	
After this call, I get following error from within the fhir-client.js on IE11:

> SCRIPT1014: Invalid character
> File: fhir-client.js, Line: 38113, Column: 3

`var state = JSON.parse(sessionStorage[params.state]); //Line: 38113`

Upon investigation, we found that sessionStorage is empty(length: 0) at this point, causing JSON.parse to throw error.

Upon further investigation on why sessionStorage is empty, I found that after below line of code (from fhir-client.js) the sessionStorage gets empty in IE11 (not in Chrome)
`window.location.href = redirect_to; // line 38571`

This whole scenario works smoothly on Chrome.

In case of any ambiguity or further queries, please feel free to ask.

**Comment by azazlhr** [comment]
Any thoughts?

**Comment by vlad-ignatov** [comment]
I'm sorry. I have a few deadlines and can't alway answer on time.

I understand your example, but it doesn't look right. In the launch page (AllScripts/Launch.html) you have a call to `FHIR.oauth2.authorize`. When that gets executed, the browser should redirect you to the auth server. You might have to select a patient, login, authorize the app etc. After all that is complete you should be redirected back to http://localhost:52447. That is where the error is happening, right?

What I'm saying is that it is normal for the sessionStorage to appear empty after a redirect to different domain. The question is, is it empty on your index page and if so - why?

Here are some questions that might help explaining the situation:
1. Is the sessionStorage empty after all those redirects?
2. What happens if you use relative path for `redirect_uri`?
3. Are there any warnings or errors in the console.
4. Do you see additional URL parameters after the redirect (like `error` or `error_description`)?
5. Try setting `FHIR.oauth2.settings.fullSessionStorageSupport = false` before the authorize call and see if that works.
6. Do you have some weird security settings in IE that can mess with the sessionStorage behavior?
7. Last but not least, do you have a specific reason to develop an app with an old (no longer maintained) version of the client library?

**Comment by azazlhr** [comment]
Thank you @vlad-ignatov  for such a comprehensive answer and excellent questions.
Before answering each question, I would like to share one more finding that I observed yesterday. I don't see this issue happening when I tested my SMART application after deploying it on a staging server (instead of localhost). So looks like IE and localhost has something to do to settle things up.

Now answering each questions:
1. Yes
2. Have to try that
3. No
4. No
5. Have to try that
6. Don't think so
7. I think moving to latest version would be major shift and could introduce some breaking changes in my application.
